software/py/bsg_manycore_link_gen.py

```diff
@@ -0,0 +1,295 @@
+
+    script = "{0} {1} :\n{2} {{".format(name, vaddr, laddr)
+
+    for sec in in_sections:
+      script += "\n  {0}({1})".format(in_objects, sec)
+
+    if boundary is not None:
+        script += "\n  PROVIDE (__{}_end = .);\n".format(name)
+        script += "  ASSERT((__{0}_end < 0x{1:08x}), " \
+                  "\"Error: {0} section exceeded limit 0x{1:08x}\");".format(name, boundary)
+
+    script += "\n  . = ALIGN(8);\n"
+    script += "\n}} {0} {1}\n\n".format(vma_redirect, lma_redirect)
+
+    return script
+
+  def script(self):
+    """
+    This generates the link script by linking in two memory regions,
+    DMEM_VMA and DRAM_LMA.
+
+    We use the combination of VMA and LMA because `.text` section VMAs and
+    dmem memory VMAs overlap with the current memory layout (both start at
+    address 0x0). This makes manycore essentially a Harvard architecture.
+    But linker doesn't allow overlapping VMAs and doesn't allow location
+    couter to move backwards. To overcome this, we use a combination of VMA
+    and LMA "memory regions" to be able to never move location counter
+    backward.
+    """
+    # Address constants:
+    #
+    # LMA (Load Memory Address)    => NPA used by loader
+    # VMA (Virtual Memory Address) => Logical address used by linker for
+    #                                 symbol resolutions
+    _DMEM_VMA_START   = 0x1000
+    _DMEM_VMA_SIZE    = 0x1000
+    _DRAM_T_LMA_START = 0x80000000
+    _DRAM_T_LMA_SIZE  = self._imem_size
+    _DRAM_D_LMA_START = 0x80000000 + self._imem_size
+    _DRAM_D_LMA_SIZE  = self._dram_size - self._imem_size
+
+    mem_regions = [
+      # Format:
+      # [<mem region name>, <access>, <start address>, <size>]
+      ['DMEM_VMA', 'rw', _DMEM_VMA_START, _DMEM_VMA_SIZE],
+      ['DRAM_T_LMA', 'rx', _DRAM_T_LMA_START, _DRAM_T_LMA_SIZE],
+      ['DRAM_D_LMA', 'rw', _DRAM_D_LMA_START, _DRAM_D_LMA_SIZE],
+      ]
+
+    section_map = [
+      # Format:
+      # <output section>: [<input sections>]
+      ['.text.dram'        , ['.crtbegin','.text','.text.startup','.text.*']],
+      ['.dmem'             , ['.dmem','.dmem.*']],
+      ['.data'             , ['.data','.data*']],
+      ['.sdata'            , ['.srodata.cst16','.srodata.cst8','.srodata.cst4',
+                              '.srodata.cst2','.srodata*','.sdata','.sdata.*',
+                              '.gnu.linkonce.s.*']],
+      ['.sbss'             , ['.sbss','.sbss.*','.gnu.linkonce.sb.*','.scommon']],
+      ['.bss'              , ['.bss','.bss*']],
+      ['.tdata'            , ['.tdata','.tdata*']],
+      ['.tbss'             , ['.tbss','.tbss*']],
+      ['.eh_frame'         , ['.eh_frame','.eh_frame*']],
+      ['.striped.data.dmem', ['.striped.data']],
+      ['.rodata.dram'      , ['.rodata','.rodata*']],
+      ['.dram'             , ['.dram','.dram.*', '.*']],
+      ]
+
+    sections = "SECTIONS {\n\n"
+
+    # DMEM sections
+    for i, m in enumerate(section_map):
+      sec = m[0]
```

Comment:
I'm trying to understand what is happening the two for loops here. It appears that the code is iterating through the section map and finding strings that only contain .dmem (hence the $). After that, I'm a little confused.

Can you help me out? I'm happy to hop on a zoom and have you walk me through it. 


Likewise, for the lower for loop with DRAM.

Reply:
Cool, I'll ping you!